transport/cluster.go

```diff
@@ -3,42 +3,113 @@ package transport
 import (
 	"fmt"
 	"log"
+	"strconv"
 	"time"
 
 	"scylla-go-driver/frame"
 	. "scylla-go-driver/frame/response"
 
+	"github.com/google/btree"
 	"go.uber.org/atomic"
 )
 
 type (
-	PeerMap = map[string]*Node
+	PeerMap         = map[string]*Node
+	DCRacksMap      = map[string]int
+	DCRepFactorsMap = map[string]repFactor
+	KSMap           = map[string]keyspace
 
 	requestChan chan struct{}
+
+	repFactor uint32
 )
 
+const BTreeDegree = 2
+
 type Cluster struct {
-	peers             atomic.Value // PeerMap
+	topology          atomic.Value // *topology
 	control           *Conn
 	cfg               ConnConfig
 	handledEvents     []frame.EventType // This will probably be moved to config.
 	knownHosts        []string
+	keyspace          string
 	refreshChan       requestChan
 	reopenControlChan requestChan
 	closeChan         requestChan
+}
+
+type topology struct {
+	peers     PeerMap
+	dcRacks   DCRacksMap
+	allNodes  []*Node
+	ring      btree.BTree
+	keyspaces KSMap
+}
 
-	// Yet to be used.
-	// ring        map[Token]*Node
-	// dataCenters map[string]*DataCenter
+type keyspace struct {
+	stg strategy
+	// TODO: Add and use attributes below
+	// tables		      map[string]table
+	// user_defined_types map[string](string, cqltype)
 }
 
-// type dataCenter struct {
-// 	nodes   []Node
-// 	rackCnt uint // Type?
-// }
+type stgClass int
+
+const (
+	networkTopology stgClass = iota
+	simple
+	local
+	other
+)
+
+type strategy struct {
+	class stgClass
+	// used in simple strategy
+	rf repFactor
+	// used in networkTopology strategy
+	dcRF DCRepFactorsMap
+	// used in other strategy
+	name string
+	data map[string]string
+}
+
+type QueryInfo struct {
+	token    *Token
+	topology *topology
+	stg      strategy
+}
+
+func (c *Cluster) newQueryInfo(t *Token) QueryInfo {
+	top := c.topology.Load().(*topology)
+	return QueryInfo{
+		token:    t,
+		topology: top,
+		stg:      top.keyspaces[c.keyspace].stg, // TODO: do we assume, that keyspace is always specified?
```

Comment:
If keyspace is not specified the token is nil. Unless you can tell the keyspace from the prepare result.

Reply:
Namely https://github.com/apache/cassandra/blob/d70f6e0d597e888b3629c6c01d05685cf79a3d12/doc/native_protocol_v4.spec#L598